### PR TITLE
Feature flag & add basic unit tests for Proxy/BlockStore UpdateLabelNamesBloom()

### PR DESCRIPTION
**Pull request by saswatamcode**
* [ ] I added CHANGELOG entry for this change.
* [ ] Change is not relevant to the end user.

## Changes

- Adds unit tests for Proxy and BlockStore bloom filter routine
- Also, fixes some edge cases and lint
- Adds feature flag for proxyStore internal label resort that defaults to true

### CODE REVIEW

**Comment by saswatamcode** [comment]
The e2e tests fail due to the docker SHA. We might need to rebase once this gets merged!

File: pkg/store/bucket.go [hunk 1]
```diff
@@ -1658,15 +1657,21 @@ func (s *BucketStore) UpdateLabelNamesBloom(ctx context.Context) error {
...
 			if err != nil {
 				return errors.Wrapf(err, "label names for block %s", b.meta.ULID)
 			}
 
+			extRes := make([]string, 0, len(b.extLset))
+			for _, l := range b.extLset {
+				extRes = append(extRes, l.Name)
+			}
+
+			res = strutil.MergeSlices(res, extRes)
```
Comment (fpetkovski) [inline pkg/store/bucket.go]: This will cause an unnecessary allocation. Let's just iterate through all extRes instead of making a new slice.

File: pkg/store/proxy_heap.go [hunk 2]
```diff
@@ -897,11 +898,15 @@ type respSet interface {
...
+func hasInternalReplicaLabels(st Client, req *storepb.SeriesRequest, internalLabelResort bool) bool {
+	if !internalLabelResort {
+		return true
+	}
+
 	bloom := st.LabelNamesBloom()
 	// Empty bloom filter capacity is 1. We fallback to eager retrieval if bloom filter
 	// is yet to be updated, as we cannot yet determine if the store has internal replica labels.
-	if bloom.Cap() <= 1 {
+	if bloom == nil || bloom.Cap() <= 1 {
```
Comment (fpetkovski) [inline pkg/store/proxy_heap.go]: Could `bloom` ever be nil?

Comment (saswatamcode) [reply pkg/store/proxy_heap.go]: Yes, when it has yet to be initialized via the first UpdateLabelNamesBloom call. 

I noticed a panic in the test here which occurs on line 904 in https://github.com/thanos-io/thanos/pull/6317 due to a nil pointer deference: https://app.circleci.com/pipelines/github/thanos-io/thanos/13661/workflows/4023a1f2-8d19-4793-98a7-232c9df35318/jobs/23598.

Comment (fpetkovski) [reply pkg/store/proxy_heap.go]: Shouldn't we have an AlwaysTrue filter in that case? If not, that could be the culprit.

Comment (saswatamcode) [reply pkg/store/proxy_heap.go]: Ah, you're right, it was initialized for promMetadata but not for others. Will do that! I'd still keep this check just to prevent potential panics like this for other uses of proxy_help

Comment (fpetkovski) [reply pkg/store/proxy_heap.go]: I don't know, IMO a bloom of `nil` should be considered a bug. Otherwise we have to add checks for nil all over the codebase.

Comment (saswatamcode) [reply pkg/store/proxy_heap.go]: I see. Will remove then.
